src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -62,6 +64,287 @@ static ShardPlacement * TupleToShardPlacement(TupleDesc tupleDesc,
+			intervalList = lappend(intervalList, shardInterval);
+		}
+	}
+
+	return intervalList;
+}
+
+
+/*
+ * GenerateSizeQueryOnMultiplePlacements generates a select size query to get
+ * size of multiple tables from the relation with distributedRelationId. Note that,
+ * different size functions supported by PG are also supported by this function. Check
+ * CitusSizeFunctionTypes for the list of supported size functions on Citus.
+ */
+StringInfo
+GenerateSizeQueryOnMultiplePlacements(Oid distributedRelationId, List *intervalsOnNode,
+									  CitusSizeFunctionTypes sizeFunctionType)
+{
+	/* build shard qualified name */
+	Oid schemaId = get_rel_namespace(distributedRelationId);
+	char *schemaName = get_namespace_name(schemaId);
+
+	StringInfo selectQuery = makeStringInfo();
+	bool isFirstShard = true;
+
+	ListCell *shardCell = NULL;
+
+	/* check whether the size of table can be obtained with citus size functions */
+	ErrorIfNotSuitableToGetSize(distributedRelationId);
+
+	foreach(shardCell, intervalsOnNode)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardCell);
+		uint64 shardId = shardInterval->shardId;
+		char *shardName = get_rel_name(distributedRelationId);
+		char *shardQualifiedName = NULL;
+		char *quotedShardName = NULL;
+		AppendShardIdToName(&shardName, shardId);
+
+		shardQualifiedName = quote_qualified_identifier(schemaName, shardName);
+		quotedShardName = quote_literal_cstr(shardQualifiedName);
+
+		if (isFirstShard)
+		{
+			appendStringInfo(selectQuery, "Select ");
+			isFirstShard = !isFirstShard;
+		}
+
+		if (sizeFunctionType == CITUS_TOTAL_RELATION_SIZE)
+		{
+			appendStringInfo(selectQuery, INNER_SHARD_TOTAL_RELATION_SIZE_QUERY,
+							 quotedShardName);
+		}
+		else if (sizeFunctionType == CITUS_TABLE_SIZE)
+		{
+			appendStringInfo(selectQuery, INNER_SHARD_TABLE_SIZE_QUERY,
+							 quotedShardName);
+		}
+		else if (sizeFunctionType == CITUS_RELATION_SIZE)
+		{
+			appendStringInfo(selectQuery, INNER_SHARD_RELATION_SIZE_QUERY,
+							 quotedShardName);
+		}
+	}
+
+	/* there is no more size function so just add 0 instead of it */
+	appendStringInfo(selectQuery, "0;");
+	return selectQuery;
+}
+
+
+/*
+ * SupportFunctionForCitusSize is a helper function for all citus size functions. It
+ * first checks whether the table is distributed and size query can be run on it. For
+ * distributed tables with streaming replications connection to each node has to be
+ * established. For reference tables, connecting to one of the node is enough.
+ */
+uint64
+SupportFunctionForCitusSize(Oid distributedRelationId, CitusSizeFunctionTypes
+							sizeFunctionType)
+{
+	DistTableCacheEntry *distributedTableCacheEntry = NULL;
+
+	List *workerNodeList = WorkerNodeList();
+	ListCell *workerNodeCell = NULL;
+
+	uint64 totalRelationSize = 0;
+
+	MultiConnection *connection = NULL;
+
+	distributedTableCacheEntry = DistributedTableCacheEntry(distributedRelationId);
+
```

Comment:
I meant to check if a table is a reference table and calculate its size accordingly here, not in the GenerateSizeQueryOnMultiplePlacements. This would save you dealing with all cases in the loop below and make the code more readable.

Also we can error out if an error occurs during query execution in first placement of reference table similar to other distributed tables.